src/client/components/robot/model.ts

```diff
@@ -1,14 +1,32 @@
+import { action } from 'mobx'
 import { observable } from 'mobx'
 
+import { Rate } from './rate'
+
+type RobotModelOpts = {
+  id: string
+  connected: boolean
+  enabled: boolean
+  name: string
+  address: string
+  port: number
+}
+
 export class RobotModel {
   @observable id: string
   @observable connected: boolean
   @observable enabled: boolean
   @observable name: string
   @observable address: string
   @observable port: number
+  @observable stats = {
```

Comment:
I'm not sure this belongs here... I think this data is too specific to live up at the top-model level.

This is network-specific, so it feels like something that should live in `/client/network/model.ts`, indicated by the fact that `/client/network/nusight_network.ts` will be updating the model anyway, so it should probably own the model data structure as well.

Can we move the model data there?

Reply:
So have a model property on `NUsightNetwork` that'll have a list of stats corresponding to the app-level `RobotModel`s?

If we do so, any ideas on how we could expose that to show in the UI? Like we do here: https://github.com/NUbots/NUsight2/blob/paye/robot-packet-rate/src/client/components/robot_selector/view.tsx#L45

Reply:
I'm thinking something like (rough pseudocode):

In `network/nusight_network.ts`:

```ts
type RobotNetworkStats = {
  packets: number
  packetsPerSecond: Rate
  bytes: number,
  bytesPerSecond: Rate
}

class NUsightNetwork {
  stats: Map<string, RobotNetworkStats>

  constructor(..., private clock: Clock) {
    this.stats = new Map()
  }

  onNUClearJoin(cb: (peer: NUClearNetPeer) => void) {
    this.nuclearnetClient.onJoin((peer) => {
      cb(peer)

      const robot = findRobot(peer)

      this.stats.set(robot.id, {
        packets: 0,
        packetsPerSecond: Rate.of({ ... }, this.clock),
        // ...
      })
    })
  }

  onNUClearMessage<T>(messageType: MessageType<T>, cb: MessageCallback<T>) {
    //...
    return this.nuclearnetClient.on(messageTypeName, (packet: NUClearNetPacket) => {
      //...
      if (robotModel) {
        this.updateStats(this.stats.get(robotModel.id), {
          packets: 1,
          bytes: buffer.length,
        })
        cb(robotModel, message)
      }
    })
  }

  updateStats(robotModel, { bytes, packets }) {
    const stats = this.stats.get(robotModel.id)
    stats.packets += packets
    stats.packetsPerSecond.update(packets)
    stats.bytes += bytes
    stats.bytesPerSecond.update(bytes)
  }
}
```

In `robot_selector/robot_label/view_model.ts`:

```ts
class RobotLabelViewModel {
  //...
  @computed stats() {
    return this.network.stats.get(this.model.id) // has access to NUsightNetwork as `network` somehow?
  }
}
```

In `robot_selector/robot_label/view.tsx`:

```tsx
<div className={style.dataCell}>
  <div className={style.dataLabel}>Packets</div>
  <div className={style.dataValue}>{
    formatSI(viewModel.stats.packets)
  }</div>
</div>
```

Probably won't work, but is it like what you had in mind?

Reply:
I recommend keeping the model separate to the `NUsightNetwork` class.

We already do this for all the other model types. So I would expect something like this:

e.g.
```ts
class RobotNetworkStatsModel {
  @observable.ref packets: number
  ...

  static of = memoize((robotModel: RobotModel) => {
    return new RobotNetworkStatsModel(robotModel)
  })
}
```

```ts
class NUsightNetwork {
  @action
  onNUClearMessage() {
    const stats = RobotNetworkStatsModel.of(robotModel)
    stats.packets++
  }
}
```

```ts
class MyComponent extends Component<{ robotModel: RobotModel }> {
  @computed
  private get stats() {
    // Note: It would probably be more correct to pass this dependency in as a prop, but that might be difficult with the current architecture of the robot selector.
    return RobotNetworkStatsModel.of(this.props.robotModel)
  }

  render() {
    return <div>{this.stats.packets}</div>
  }
}
```

Reply:
This is much simpler, thanks!

Reply:
Implemented in https://github.com/NUbots/NUsight2/pull/269/commits/c9e872156322db2df91be7b947247c2ebbb24f94